lib/pushmi_pullyu/aip/downloader.rb

```diff
@@ -1,215 +1,184 @@
+require 'fileutils'
+require 'ostruct'
+require 'rdf'
 require 'rdf/n3'
-require 'pushmi_pullyu/aip'
-require 'pushmi_pullyu/aip/solr_fetcher'
-require 'pushmi_pullyu/aip/fedora_fetcher'
 
 # Download all of the metadata/datastreams and associated data
 # related to an object
-
 class PushmiPullyu::AIP::Downloader
 
-  FILENAME_PREDICATE = 'info:fedora/fedora-system:def/model#downloadFilename'.freeze
+  class NoContentFilename < StandardError; end
 
   def initialize(noid)
     @noid = noid
-    @fedora_fetcher = PushmiPullyu::AIP::FedoraFetcher.new(@noid)
-    PushmiPullyu::AIP.validate(noid)
   end
 
   def run
     make_directories
+
     PushmiPullyu.logger.info("#{@noid}: Retreiving data from Fedora ...")
 
     [:main_object, :fixity, :content_datastream_metadata, :versions, :thumbnail,
      :characterization, :fedora3foxml, :fedora3foxml_metadata].each do |item|
       path_spec = aip_paths[item]
-      download_and_log(path_spec)
+      download_and_log(path_spec, PushmiPullyu::AIP::FedoraFetcher.new(@noid))
     end
 
     # Need content filename from metadata
-    download_and_log(aip_paths.content, local_path: content_filename)
+    path_spec = OpenStruct.new(
+      remote: '/content',
+      local: content_filename, # lookup filename derived from metadata
+      optional: false
+    )
+    download_and_log(path_spec, PushmiPullyu::AIP::FedoraFetcher.new(@noid))
 
     download_permissions
-
-    # Return directory name
-    PushmiPullyu::AIP.aip_directory(@noid)
   end
 
   private
 
-  def download_and_log(path_spec, local_path: nil, fedora_fetcher: nil)
-    # Fetcher is either for current object, or for a permission object
-    fedora_fetcher ||= @fedora_fetcher
-
-    # Sometimes we don't know filename in advance, so we use local_path ...
-    output_file = full_local_path(local_path || path_spec.local)
-    url = fedora_fetcher.object_url(path_spec.remote)
+  def download_and_log(path_spec, fedora_fetcher)
```

Comment:
Simplify this. There is only one time that the local_path is nil, which is when we need to get the content_filename. This can be done smarter and fix by including it in the path_spec which we are passing in. Also there is only one time we are passing in the fedora_fetcher which is for permissions. Instead of doing a nil check and falling back to the default one, let's just pass it in every time. Much simpler API and no more nil checks everywhere.